### ISSUE TITLE
Why isn't SampledWithoutReplacementDpEvent used instead of PoissonSampledDpEvent for DP-SGD epsilon calculation?

### ISSUE THREAD

**Issue by BjarnePfitzner** [tensorflow/privacy#483]
In the readme of the tutorials, when calculating the privacy spending epsilon of the DP-SGD training, the accountant receives a DP event composed of multiple `dp_event.PoissonSampledDpEvent`s (see snippet below). https://github.com/tensorflow/privacy/blob/a4bdb05b623a59de3b849e0031b61edbf5a305dc/tutorials/walkthrough/README.md?plain=1#L411-L417

Can anyone explain why we don't use `dp_event.SampledWithoutReplacementDpEvent`? When looping over the training set, we actually sample batches without replacement, are we not?

**Comment by galenmandrew** [comment]
First of all, it is important to note that neither `PoissonSampledDpEvent` nor `SampledWithoutReplacementDpEvent` accurately describe the training procedure of looping over the (perhaps shuffled) dataset and forming batches. `SampledWithoutReplacementDpEvent` would be correct if you were sampling fixed-sized batches uniformly at random and independently across rounds. There is no known tight analysis of the shuffle and iterate style training that is most commonly used.

So the choice between `PoissonSampledDpEvent` nor `SampledWithoutReplacementDpEvent` comes down to other considerations. The main argument in favor of  `PoissonSampledDpEvent` is that the privacy guarantee you derive from it is under the "add/remove-one" notion of dataset adjacency.  `SampledWithoutReplacementDpEvent` gives you a guarantee under "replace-one" adjacency, which it is believed produces epsilon values about a factor of two larger.

For a formally correct guarantee under "add/remove-one" adjacency, in the case where each example is visited at most once, you could analyze a single (not composed) application of the unamplified Gaussian mechanism, corresponding to the scenario where the adversary knows in which round the targeted example appears.

For a better estimate of the true privacy loss, you could apply empirical privacy estimation techniques.

**Comment by BjarnePfitzner** [comment]
Thank you for your detailed explanation! That clarifies it :)
